Data/SearchEngine/Autosuggest.hs

```diff
@@ -26,6 +29,51 @@ import qualified Data.IntSet as IntSet
 import qualified Data.Vector.Unboxed as Vec
 
 
+-- | Given an incomplete prefix query, find the set of documents that match
+-- possible completions of that query.  This should be less computationally
+-- expensive than 'queryAutosuggest' as it does not do any ranking of documents.
+-- However, it does not apply the pre-filter or post-filter limits, and the list
+-- may be large when the query terms occur in many documents.  The order of
+-- returned keys is unspecified.
+queryAutosuggestMatchingDocuments :: (Ix field, Bounded field, Ord key) =>
+                                     SearchEngine doc key field feature ->
+                                     [Term] -> Term -> [key]
+queryAutosuggestMatchingDocuments se@SearchEngine{searchIndex} precedingTerms partialTerm =
+    let (_, ds, _, _) = process se precedingTerms partialTerm
+    in map (SI.getDocKey searchIndex) (DocIdSet.toList ds)
+
+-- | Given an incomplete prefix query, return a predicate that indicates whether
+-- a key is in the set of documents that match possible completions of that
+-- query.  This is equivalent to calling 'queryAutosuggestMatchingDocuments' and
+-- testing whether the key is in the list, but should be more efficient.
+--
+-- This does not apply the pre-filter or post-filter limits.
+queryAutosuggestPredicate :: (Ix field, Bounded field, Ord key) =>
+                             SearchEngine doc key field feature ->
+                             [Term] -> Term -> (key -> Bool)
+queryAutosuggestPredicate se@SearchEngine{searchIndex} precedingTerms partialTerm =
+    let (_, ds, _, _) = process se precedingTerms partialTerm
+    in (\ key -> maybe False (flip DocIdSet.member ds) (SI.getKeyDocId searchIndex key))
+
+
+process :: (Ix field, Bounded field) =>
+           SearchEngine doc key field feature ->
+           [Term] -> Term -> ([TermId], DocIdSet, [(TermId, DocIdSet)], [TermId])
+process se precedingTerms partialTerm =
```

Comment:
I think it will be clearer to read if you don't pull `process` out of `queryAutosuggest` at all, and just use `mkAutosuggestQuery` and `processAutosuggestQuery` directly within the two new functions, given that you only use one of the elements of the result tuple anyway.

The "step" naming convention makes sense in `queryAutosuggest` since it has a dozen steps, and it makes things less clear there to pull out two of those steps into a separate function. Each step itself is just a wrapper function. They're only there to clarify the structure of the thing.

Reply:
Good idea, thanks. I've changed this and it is indeed simpler.